Excerpted from <https://git.savannah.gnu.org/cgit/bash.git/commit/?h=devel&id=055a4552c901e43e6a6007f5bc664a33698dbf85>.

Original author: Chet Ramey <chet.ramey@case.edu>

--- variables.c
+++ variables.c
@@ -5413,7 +5413,9 @@ pop_var_context ()
   vcxt = shell_variables;
   if (vc_isfuncenv (vcxt) == 0)
     {
-      internal_error (_("pop_var_context: head of shell_variables not a function context"));
+      /* If we haven't flushed all of the local contexts already, flag an error */
+      if (shell_variables != global_variables || variable_context > 0)
+	internal_error (_("pop_var_context: head of shell_variables not a function context"));
       return;
     }
 
